README.md
## lsc_ros2_driver (humble) - 1.0.0-2

The packages in the `lsc_ros2_driver` repository were released into the `humble` distro by running `/usr/bin/bloom-release --new-track --rosdistro humble --track humble lsc_ros2_driver` on `Fri, 16 Dec 2022 00:53:30 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-1`
- new version: `1.0.0-2`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-10

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Fri, 16 Dec 2022 00:46:06 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-9`
- new version: `1.0.0-10`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-9

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Fri, 16 Dec 2022 00:23:18 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-8`
- new version: `1.0.0-9`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (humble) - 1.0.0-1

The packages in the `lsc_ros2_driver` repository were released into the `humble` distro by running `/usr/bin/bloom-release --new-track --rosdistro humble --track humble lsc_ros2_driver` on `Fri, 16 Dec 2022 00:10:16 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.0-1`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-8

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Fri, 16 Dec 2022 00:04:41 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-7`
- new version: `1.0.0-8`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-7

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Fri, 16 Dec 2022 00:03:16 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-6`
- new version: `1.0.0-7`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-6

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Thu, 15 Dec 2022 23:52:18 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-5`
- new version: `1.0.0-6`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-5

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Thu, 15 Dec 2022 23:42:38 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-4`
- new version: `1.0.0-5`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-4

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Thu, 15 Dec 2022 23:39:00 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-3`
- new version: `1.0.0-4`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-3

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Thu, 15 Dec 2022 23:23:28 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-2`
- new version: `1.0.0-3`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-2

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Thu, 15 Dec 2022 23:16:59 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `1.0.0-1`
- new version: `1.0.0-2`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`


## lsc_ros2_driver (foxy) - 1.0.0-1

The packages in the `lsc_ros2_driver` repository were released into the `foxy` distro by running `/usr/bin/bloom-release --new-track --rosdistro foxy --track foxy lsc_ros2_driver` on `Thu, 15 Dec 2022 09:19:53 -0000`

The `lsc_ros2_driver` package was released.

Version of package(s) in repository `lsc_ros2_driver`:

- upstream repository: https://github.com/AutonicsLiDAR/lsc_ros2_driver.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.0-1`

Versions of tools used:

- bloom version: `0.11.2`
- catkin_pkg version: `0.5.2`
- rosdep version: `0.22.1`
- rosdistro version: `0.9.0`
- vcstools version: `0.1.42`




CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package lsc_ros2_driver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (2022-12-14)
-----------
* Initial release


src/socket.cpp
/*
* Copyright (c) 2022, Autonics Co.,Ltd.
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*
*   * Neither the name of the Autonics Co.,Ltd nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/

#include <iostream>

#ifndef _MSC_VER
#include <unistd.h>
#include <sys/poll.h>
#endif

#include <string.h>
#include <errno.h>

#include <thread>

#include "socket.hpp"


tcpComm::tcpComm()
{
    m_server_sock_ = 0;
    m_connected_ = false;
    rcv_timeout_ = false;
    rcv_error_ = false;
    thread_running_ = false;
}

tcpComm::~tcpComm()
{
    CommDisconnect();
}


void tcpComm::putBufToMsg(unsigned char* buf, uint16_t size)
{
    std::vector<unsigned char> vec;
    vec.reserve(size);

    for(int i = 0; i < size; i++)
    {
        vec.push_back(*(buf+i));
    }
    recvQueue.push(vec);
}


#ifdef _MSC_VER
SOCKET tcpComm::getServerSocket(void)
#else
int tcpComm::getServerSocket(void)
#endif
{
    return m_server_sock_;
}


bool tcpComm::IsConnected(void)
{
    return m_connected_;
}

bool tcpComm::getthreadRunning(void)
{
    return thread_running_;
}


bool tcpComm::getRcvTimeout(void)
{
    return rcv_timeout_;
}

bool tcpComm::getRcvError(void)
{
    return rcv_error_;
}

void tcpComm::setConnected(bool flag)
{
    m_connected_ = flag;
}

void tcpComm::setThreadRunning(bool flag)
{
    thread_running_ = flag;
}

void tcpComm::setRcvTimeout(bool flag)
{
    rcv_timeout_ = flag;
}

void tcpComm::setRcvError(bool flag)
{
    rcv_error_ = flag;
}


void* readCallback(void* arg)
{
    int32_t read_cnt = 0;
    uint16_t packet_size = 0, total_byte = 0, offset = 0;
    int ret = 0;
    unsigned char r_buffer[16384], buffer[16384], cp_temp[16];
    tcpComm* sock = (tcpComm*) arg;

#ifdef _MSC_VER
    struct pollfd fd[1];
    fd[0].fd = sock->getServerSocket();
    fd[0].events = POLLIN;
#else
    struct pollfd fd;
    fd.fd = sock->getServerSocket();
    fd.events = POLLIN;
#endif

    while(sock->getthreadRunning())
    {
        read_cnt = 0;
        offset = 0;
#ifdef _MSC_VER
        ret = WSAPoll(fd, 1, 1000);
#else
        ret = poll(&fd, 1, 1000);
#endif
        if(ret < 0)
        {
            sock->setRcvError(true);
            std::cout << "polling error" << std::endl;
        }
        else /// ret >= 0
        {
            sock->setRcvError(false);

            read_cnt = sock->CommRead(r_buffer, sizeof(r_buffer));

            if(ret == 0)
            {
                sock->setRcvTimeout(true);
            #ifdef PRINT_DEBUG
                std::cout << "polling timeout" << std::endl;
            #endif
            }

            if(read_cnt > 0)
            {
                sock->setRcvTimeout(false);

                /// Search packet size
                if(packet_size == 0)
                {
                    for(uint16_t i = 0; i < read_cnt; i++)
                    {
                        if(r_buffer[i] == STX)
                        {
                            for(int j = 0; j < 4; j++)
                            {
                                cp_temp[j] = r_buffer[i+j+1];
                            }
                            packet_size = (uint16_t) strtoul((const char*) cp_temp, NULL, 16);
                            break;
                        }
                    }
                }

                /// Search ETX
                for(uint16_t i = 0; i < read_cnt - 1; i++)
                {
                    if(r_buffer[read_cnt - 1 - i] == ETX)
                    {
                        read_cnt -= i;
                        break;
                    }
                }

                if(packet_size > total_byte)
                {
                    memcpy(&buffer[total_byte], &r_buffer[offset], read_cnt - offset);
                    total_byte += read_cnt - offset;
                }
                else if(packet_size < total_byte)
                {
                    if(buffer[0] == 0x02 && buffer[packet_size] == 0x03)
                    {
                        total_byte = packet_size;
                    }
                    else
                    {
                        memset(buffer, 0x00, sizeof(buffer));
                        packet_size = 0;
                        total_byte = 0;
                        offset = 0;
                    }
                }
                else
                {

                }

                if(packet_size !=0 && total_byte != 0 && packet_size == total_byte)
                {
                    if(buffer[total_byte - 1] == 0x03)
                    {
                        sock->putBufToMsg(buffer, total_byte);
                        memset(buffer, 0x00, sizeof(buffer));
                        packet_size = 0;
                        total_byte = 0;
                    }
                }
                else
                {

                }

                memset(r_buffer, 0x00, sizeof(r_buffer));
            }
            else if(read_cnt < 0)    /// disconnect
            {
                std::cout << "server is disconnected" << std::endl;
                sock->setRcvError(false);
                sock->setConnected(false);
                sock->setThreadRunning(false);
            }
            else /// receive error
            {
                if(sock->CommWrite(r_buffer, sizeof(r_buffer)))
                {
                    std::cout << "checked connection, server is disconnected" << std::endl;
                    sock->setConnected(false);
                    sock->setThreadRunning(false);
                }
            }
        }
    }

    return 0;
}


void tcpComm::CommDisconnect()
{
    uint16_t bk_cnt = 0;
    thread_running_ = false;

    if(thread_read_.joinable())
    {
        thread_read_.join();
    }

#ifdef _MSC_VER
    closesocket(m_server_sock_);
#else
    while(close(m_server_sock_) !=0 || bk_cnt >= 500)
    {
        bk_cnt++;
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

#endif
    if(m_connected_ == true)
    {
        m_connected_ = false;
        std::cout << "disconnect server" << std::endl;
    }
}


int tcpComm::CommConnect(std::string addr, uint16_t port)
{
    int m_server_addr_size = 0;
    timeval tv;
    addr_st_ = addr;
    port_num_ = port;

    m_server_sock_ = socket(PF_INET, SOCK_STREAM, 0);

    if(m_server_sock_ < 0)
    {
        std::cout << "faild to create client socket" << std::endl;
        return -1;
    }

    tv.tv_sec = RS_TIMEOUT_S;
    tv.tv_usec = RS_TIMEOUT_NS;

    if(setsockopt(m_server_sock_, SOL_SOCKET, SO_RCVTIMEO, (char*)&tv, sizeof(tv)) < 0)
    {
        std::cout << "failed rcvtimeo setsockopt" << std::endl;
    }
    if(setsockopt(m_server_sock_, SOL_SOCKET, SO_SNDTIMEO, (char*)&tv, sizeof(tv)) < 0)
    {
        std::cout << "failed sndtimeo setsockopt" << std::endl;
    }

    memset(&m_server_addr_, 0x00, sizeof(m_server_addr_));

    m_server_addr_.sin_family = AF_INET;
#ifdef _MSC_VER
    m_server_addr_.sin_addr.s_addr = inet_addr(addr.c_str());
#else
    m_server_addr_.sin_addr.s_addr = inet_addr(addr.begin().base());
#endif
    m_server_addr_.sin_port = htons(port);

    m_server_addr_size = sizeof(m_server_addr_);

    if(connect(m_server_sock_, (const sockaddr*)&m_server_addr_, m_server_addr_size) < 0)
    {
        std::cout << "failed client connect, "<< errno << " " << strerror(errno) << std::endl;
        return -1;
    }
    else
    {
        m_connected_ = true;
        std::cout << "connected" << std::endl;
    }

    try
    {
        thread_read_ = std::thread(readCallback, this);
        thread_running_ = true;
        std::cout << "thread created" << std::endl;
    }
    catch(const std::exception& e)
    {
        std::cerr << e.what() << '\n';
        std::cout << "thread create failed" << std::endl;
        return -1;
    }

    return 0;
}


int tcpComm::CommRead(unsigned char* buffer, int buf_size)
{
    int read_size = 0;

    if(m_connected_ == true)
    {
        read_size = recv(m_server_sock_, (char *)buffer, buf_size, 0);
        if(read_size < 0)
        {
            std::cout << "client read failed, " << errno << " " << strerror(errno) << std::endl;
            if(errno == 104)    /// connect reset
            {
                thread_running_ = false;
                m_connected_ = false;
                read_size = 0;
            }
            else if(errno == 11)    /// no message received
            {
                read_size = 0;
            }
        }
        else if(read_size == 0)
        {
            std::cout << "server socket has closed" << std::endl;
        }
    }
    else
    {
        std::cout << "read failed, server is disconnected" << std::endl;
        read_size = -1;
    }

    return read_size;
}


int tcpComm::CommWrite(unsigned char* buffer, int buf_size)
{
    int write_size = 0;

    if(m_connected_ == true)
    {
        write_size = send(m_server_sock_, (const char *)buffer, buf_size, 0);
        if(write_size < 0)
        {
            std::cout << "client write failed, " << errno << " " << strerror(errno) << std::endl;
            if(errno == 104)
            {
                thread_running_ = false;
                m_connected_ = false;
                write_size = 0;
            }
        }
        else if(write_size == 0)
        {
            std::cout << "no byte sent" << std::endl;
        }
    }
    else
    {
        std::cout << "write failed, server is disconnected" << std::endl;
        write_size = -1;
    }

    return write_size;
}



launch/lsc_ros2_driver_launch.py
import os
from ament_index_python.packages import get_package_share_directory

from launch import LaunchDescription
from launch.substitutions import LaunchConfiguration
from launch.actions import DeclareLaunchArgument
from launch_ros.actions import Node
from launch.actions import LogInfo


def generate_launch_description():
  addr = LaunchConfiguration('addr', default='192.168.0.1')
  port = LaunchConfiguration('port', default='8000')
  frame_id = LaunchConfiguration('frame_id', default="laser")
  range_min = LaunchConfiguration('range_min', default=0.05)
  range_max = LaunchConfiguration('range_max', default=25.0)
  intensities = LaunchConfiguration('intensities', default='true')

  pub_topic = LaunchConfiguration('pub_topic', default="scan")
  diagnostics_tolerance = LaunchConfiguration('diagnostics_tolerance', default=0.1)
  diagnostics_windows_time = LaunchConfiguration('diagnostics_windows_time', default=1)


  rviz_config_dir = os.path.join(
    get_package_share_directory('lsc_ros2_driver'),
    'rviz',
    'lsc_ros2_driver.rviz'
  )

  return LaunchDescription([

    DeclareLaunchArgument(
      'addr',
      default_value=addr,
      description='server address to connect'
    ),

    DeclareLaunchArgument(
      'port',
      default_value=port,
      description='server port opened'
    ),

    DeclareLaunchArgument(
      'frame_id',
      default_value=frame_id,
      description='topic frame_id'
    ),

    DeclareLaunchArgument(
      'range_min',
      default_value=range_min,
      description='minimum range value of device'
    ),

    DeclareLaunchArgument(
      'range_max',
      default_value=range_max,
      description='maximum range value of device'
    ),

    DeclareLaunchArgument(
      'intensities',
      default_value=intensities,
      description='intensities enable'
    ),

    DeclareLaunchArgument(
      'pub_topic',
      default_value=pub_topic,
      description='topic name'
    ),

    DeclareLaunchArgument(
      'diagnostics_tolerance',
      default_value=diagnostics_tolerance,
      description='set diagnostics_tolerance value'
    ),

    DeclareLaunchArgument(
      'diagnostics_windows_time',
      default_value=diagnostics_windows_time,
      description='set diagnostics_windows_time value'
    ),

    Node(
      package='lsc_ros2_driver',
      executable='autonics_lsc_lidar',
      name='autonics_lsc_lidar',
      parameters=[{
        'addr': addr,
        'port': port,
        'frame_id' : frame_id,
        'range_min': range_min,
        'range_max': range_max,
        'intensities': intensities,
        'pub_topic' : pub_topic,
        'diagnostics_tolerance' : diagnostics_tolerance,
        'diagnostics_windows_time' : diagnostics_windows_time
      }],
      output='screen'),

    Node(
      package='rviz2',
      executable='rviz2',
      name='rviz2',
      arguments=['-d' + rviz_config_dir],
      output='screen'),
  ])


include/lsc_ros2_driver/parser.hpp
/*
* Copyright (c) 2022, Autonics Co.,Ltd.
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*
*   * Neither the name of the Autonics Co.,Ltd nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/


#ifndef PARSER_HPP
#define PARSER_HPP

#include <vector>

// #define PRINT_PARSED_DATA

#define PI  3.14159265358979323846
#define DEG2RAD PI / 180

#define NUM_OF_CMD_TYPE 4
#define NUM_OF_CMD 7

#define NUMBER_OF_PARAM 13

enum CmdListNum
{
    NONE = 0,
    SET_ACCESS_LEVEL,
    SENSOR_SCAN_INFO,
    SENSOR_START,
    SENSOR_STOP,
    SCAN_DATA,
    FIRST_CONNECT_DUMMY_SEND
};


struct ScanInfo
{
    uint16_t fw_ver;
    std::string model_name;
};

struct ScanMea
{
    uint16_t scan_counter;
    uint16_t scan_freq;
    uint16_t meas_freq;
    int32_t angle_begin;
    uint16_t angle_resol;
    uint16_t amnt_of_data;
    uint16_t active_field_num;

    std::vector<float> ranges;
    std::vector<float> rssi;
};

struct Lsc_t
{
    ScanMea scan_mea;
    ScanInfo scan_info;
};


class Parser
{
  public:
    virtual void parsingMsg(std::vector<unsigned char> raw_msg, std::shared_ptr<Lsc_t> parsed_data) = 0;
    virtual std::string getResponse() = 0;
};

class AsciiParser : public Parser
{
  public:
    void parsingMsg(std::vector<unsigned char> raw_msg, std::shared_ptr<Lsc_t> parsed_data);
    std::string getResponse();

  private:
    std::string command_;
};


#endif


src/parser.cpp
/*
* Copyright (c) 2022, Autonics Co.,Ltd.
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*
*   * Neither the name of the Autonics Co.,Ltd nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/


#include <queue>
#include <cstring>
#include <iostream>

#include <memory>

#include "parser.hpp"

#include <chrono>


const char* cmd_type_list[NUM_OF_CMD_TYPE] =
{"sMC",
"sMA",
"sRC",
"sRA",
};

const char* cmd_list[NUM_OF_CMD] =
{"None",
"SetAccessLevel",
"SensorScanInfo",
"SensorStart",
"SensorStop",
"ScanData",
"FirstConnectDummySend"
};

std::string AsciiParser::getResponse()
{
    return command_;
}


void AsciiParser::parsingMsg(std::vector<unsigned char> raw_msg, std::shared_ptr<Lsc_t> parsed_data)
{
    uint16_t index= 0;
 #ifdef PRINT_PARSED_DATA
    int cmd_type = 0;
#endif
    int cmd = 0;
    std::vector<char *> field;
    char* ptr;


    if(raw_msg[0] == 0x02)
    {
    #ifdef PRINT_PARSED_DATA
        std::cout << "raw_msg : ";
        for(uint16_t i = 0; i < raw_msg.size(); i++)
        {
            std::cout << raw_msg[i];
        }
        std::cout << "" << std::endl;
    #endif


        ptr = strtok((char*) &raw_msg[1], ",");

        if(raw_msg[strtoul(ptr,NULL, 16) - 1] == 0x03)
        {
            raw_msg[strtoul(ptr, NULL, 16) - 1] = '\0';
        }

        while(ptr != NULL)
        {
            field.push_back(ptr);
            ptr = strtok(NULL, ",");
        }
        index++;

    #ifdef PRINT_PARSED_DATA
        for(int i = 0; i < NUM_OF_CMD_TYPE; i++)
        {
            int temp_ret = strcmp(field[index], cmd_type_list[i]);

            if(temp_ret == 0)
            {
                cmd_type = i;
                break;
            }
        }
    #endif
        index++;

        command_ = field[index];

        for(int i = 0; i < NUM_OF_CMD; i++)
        {
            if(strcmp(field[index], cmd_list[i]) == 0)
            {
                cmd = i;
                break;
            }
        }
        index++;


        switch(cmd)
        {
            case SENSOR_SCAN_INFO :
                parsed_data->scan_info.fw_ver = (uint16_t)strtoul(field[8], NULL, 16);
                parsed_data->scan_info.model_name = field[13];

                #ifdef PRINT_PARSED_DATA
                std::cout << "Command : " << cmd_list[cmd] << std::endl;
                std::cout << "fw_ver : " << parsed_data->scan_info.fw_ver << std::endl;
                std::cout << "Model_name : " << parsed_data->scan_info.model_name << std::endl;
                #endif

                break;

            case SCAN_DATA :


                for(int i = 0; i < NUMBER_OF_PARAM; i++)
                {
                    ptr = strtok(NULL, ",");
                    field.push_back(ptr);
                }

                parsed_data->scan_mea.scan_counter = (uint16_t)strtoul(field[7], NULL, 16);
                parsed_data->scan_mea.scan_freq = (uint16_t)strtoul(field[10], NULL, 16);
                parsed_data->scan_mea.meas_freq = (uint16_t)strtoul(field[11], NULL, 16);
                parsed_data->scan_mea.angle_begin = (int32_t)strtoul(field[12], NULL, 16);
                parsed_data->scan_mea.angle_resol = (uint16_t)strtoul(field[13], NULL, 16);
                parsed_data->scan_mea.amnt_of_data = (uint16_t)strtoul(field[14], NULL, 16);
                index = 16;

                #ifdef PRINT_PARSED_DATA
                std::cout << "Cmd type : " << cmd_type_list[cmd_type] << std::endl;
                std::cout << "Command : " << cmd_list[cmd] << std::endl;
                std::cout << "Scan counter : " << parsed_data->scan_mea.scan_counter << std::endl;
                std::cout << "Scan frequency : " << std::dec << parsed_data->scan_mea.scan_freq << std::endl;
                std::cout << "Measuring frequency : " << parsed_data->scan_mea.meas_freq << std::endl;
                std::cout << "Angle begin : " << parsed_data->scan_mea.angle_begin << std::endl;
                std::cout << "Angle resolution : " << parsed_data->scan_mea.angle_resol << std::endl;
                std::cout << "Amount of data : " << parsed_data->scan_mea.amnt_of_data << std::endl;
                #endif


                parsed_data->scan_mea.ranges.clear();
                if(strcmp(field[index++], "DIST1") == 0)
                {
                    parsed_data->scan_mea.ranges.reserve(parsed_data->scan_mea.amnt_of_data);
                    for(uint16_t i = 0; i < parsed_data->scan_mea.amnt_of_data; i++)
                    {
                        parsed_data->scan_mea.ranges.emplace_back((float) (strtoul(field[index++], NULL, 16) / 1000.0));
                    }

                    #ifdef PRINT_PARSED_DATA
                    std::cout << "ranges : " << std::endl;
                    for(uint16_t i = 0; i < parsed_data->scan_mea.ranges.size(); i++)
                    {
                        std::cout << parsed_data->scan_mea.ranges[i] << " ";
                    }
                    std::cout << "" << std::endl;
                    #endif
                }

                parsed_data->scan_mea.rssi.clear();
                if(field[index] != field.back())
                {
                    if(strcmp(field[index++], "RSSI1") == 0 )
                    {
                        parsed_data->scan_mea.rssi.reserve(parsed_data->scan_mea.amnt_of_data);
                        for(uint16_t i = 0; i < parsed_data->scan_mea.amnt_of_data; i++)
                        {
                            parsed_data->scan_mea.rssi.emplace_back((float) strtoul(field[index++], NULL, 16));
                        }

                        #ifdef PRINT_PARSED_DATA
                        std::cout << "rssi : " << std::endl;
                        for(uint16_t i = 0; i < parsed_data->scan_mea.rssi.size(); i++)
                        {
                            std::cout << parsed_data->scan_mea.rssi[i] << " ";
                        }
                        std::cout << "" << std::endl;
                        #endif
                    }
                }
                break;

            case SENSOR_START :
            case SENSOR_STOP :
            case SET_ACCESS_LEVEL :
            case FIRST_CONNECT_DUMMY_SEND :
                for(std::size_t i = 2; i < field.size(); i++)
                {
                    std::cout << field[i] << " ";
                }
                std::cout << "" << std::endl;
                break;

            default :
                for(std::size_t i = 2; i < field.size(); i++)
                {
                    std::cout << field[i] << " ";
                }
                std::cout << "" << std::endl;
                break;
        }

    raw_msg.clear();
    }
}
